### PR TITLE
Support deployable instance

### PR DESCRIPTION
**Pull request by droot**
This PR adds support for deployable instance in `kpt`. Key changes

1. `kpt pkg get --for-deployment` to indicate deployable instance
2. `deployable` hook as library
3. builtin function for generating package context

### CODE REVIEW

File: internal/hook/executor.go [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+	"fmt"
+	"io"
+
+	"github.com/GoogleContainerTools/kpt/internal/fnruntime"
+	"github.com/GoogleContainerTools/kpt/internal/types"
+	"github.com/GoogleContainerTools/kpt/internal/util/cmdutil"
+	fnresult "github.com/GoogleContainerTools/kpt/pkg/api/fnresult/v1"
+	kptfilev1 "github.com/GoogleContainerTools/kpt/pkg/api/kptfile/v1"
+	"github.com/GoogleContainerTools/kpt/pkg/fn"
+	"sigs.k8s.io/kustomize/api/filesys"
```
Comment (martinmaly) [inline internal/hook/executor.go:28]: use sigs.k8s.io/kustomize/kyaml/filesys

**Review by martinmaly — CHANGES_REQUESTED** [review]
Most nits although the test failures seen real and should be fixed.

File: internal/builtins/pkg_context_test.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+	exp    string
+	expErr error
+}
+
+func TestPkgContextGenerator(t *testing.T) {
+
+	tests := []test{
+		{
+			name: "pkg context should succeed on a non-nested package",
+			in: `apiVersion: config.kubernetes.io/v1
```
[the remark this answers is not in the document]
Comment (droot) [reply internal/builtins/pkg_context_test.go]: Done :)

File: internal/hook/executor.go (hunk 1 above)
Comment (droot) [reply internal/hook/executor.go]: Aah... goimports :)

File: internal/util/get/get.go [hunk 3]
```diff
@@ -52,6 +53,10 @@ type Command struct {
 	// Name is the name to give the package.  Defaults to the destination.
 	Name string
 
+	// IsDeploymentInstance indicates if the package is forked for deployment.
+	// If forked package has defined deploy hooks, those will be executed post fork.
+	IsDeploymentInstance bool
```
[the remark this answers is not in the document]
Comment (droot) [reply internal/util/get/get.go:58]: Good point. I have considered that in the long term proposal where one can invoke these hooks `kpt fn evalhook`. For the limited scope, one should be able to invoke these builtins function using `kpt fn eval` (I will have to adjust `kpt fn eval` to invoke builtin functions).

File: internal/hook/executor.go [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+			}
+			e.dockerCheckDone = true
+		}
+		runner, err = fnruntime.NewRunner(ctx,
+			e.FileSystem,
+			&fn,
+			types.UniquePath(e.PkgPath),
+			e.fnResults,
+			e.ImagePullPolicy,
+			false,
```
[the remark this answers is not in the document]
Comment (droot) [reply internal/hook/executor.go]: fixed.

**Review by droot — COMMENTED** [review]
Thanks @martinmaly Addressed the feedback in the latest commit. PTAL.

File: internal/builtins/pkg_context.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+						Severity: framework.Error,
+					},
+				}
+				return resourceList.Results
+			}
+			contextResources = append(contextResources, pkgContext)
+		}
+	}
+
+	for _, resource := range contextResources {
```
[the remark this answers is not in the document]
Comment (martinmaly) [reply internal/builtins/pkg_context.go:76]: I'd be happy to chat about Porch experience with virtual fs vs. resource list vs. simple representation such as map string->string (file path -> contents).

**Review by martinmaly — APPROVED** [review]

File: internal/util/get/get.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (martinmaly) [reply internal/util/get/get.go:58]: sg. something to keep in mind and do later. Thank you!

File: internal/builtins/pkg_context.go [hunk 6]
```diff
@@ -88,19 +89,16 @@ func (pc *PackageContextGenerator) Process(resourceList *framework.ResourceList)
...
   annotations:
     config.kubernetes.io/local-config: "true"
-    internal.config.kubernetes.io/path: 'package-context.yaml'
 data: {}
-`)
-	if err := cm.SetName(pkgContextName); err != nil {
-		return nil, err
-	}
+`, pkgContextName))
+
```
Comment (yuwenma) [inline internal/builtins/pkg_context.go:101]: qq, shall we use SetName to update the name field?

File: internal/builtins/pkg_context.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+const pkgContextFile = "package-context.yaml"
+const pkgContextName = "kptfile.kpt.dev"
+
+// PackageContextGenerator is a built-in KRM function that generates
+// a KRM object that contains package context information that can be
+// used by functions such as `set-namespace` to customize package with
+// minimal configuration.
+type PackageContextGenerator struct{}
+
+func (pc *PackageContextGenerator) Run(r io.Reader, w io.Writer) error {
```
Comment (yuwenma) [inline internal/builtins/pkg_context.go:40]: nit: Docstring here and below.

File: internal/builtins/pkg_context.go [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+
+func isKptfile(resource *yaml.RNode) bool {
+	gvk := resid.GvkFromNode(resource)
+	return gvk.Kind == kptfilev1.KptFileName &&
+		gvk.ApiVersion() == kptfilev1.TypeMeta.APIVersion
+}
+
+func isPkgContext(resource *yaml.RNode) bool {
+	gvk := resid.GvkFromNode(resource)
+	return resource.GetName() == pkgContextName && gvk.Kind == "ConfigMap" && gvk.ApiVersion() == "v1"
```
Comment (yuwenma) [inline internal/builtins/pkg_context.go:129]: nit:
var pkgGVK = resid.NewGvk("kptfile.kpt.dev", "v1", "ConfigMap")

func isPkgContext() {
  gvk := resid.GvkFromNode(resource)
  return gvk.Equals(pkgGVK)
} # Or maybe don't need the func?

File: internal/builtins/pkg_context.go [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+	// This loop does the following:
+	// - Filters out package context resources from the input resources
+	// - Generates a package context resource for each kpt package (i.e Kptfile)
+	for _, resource := range resourceList.Items {
+		if isPkgContext(resource) {
+			// drop existing package context resources
+			continue
+		}
+		updatedResources = append(updatedResources, resource)
+		if isKptfile(resource) {
```
Comment (yuwenma) [inline internal/builtins/pkg_context.go:61]: would it be more readable if 

```go
var pkgContextGVK = resid.NewGvk("kptfile.kpt.dev", "v1", "ConfigMap")
var KptfileGVK =  resid.NewGvk(resid.kptfilev1.TypeMeta.APIVersion(kptfilev1.TypeMeta.APIVersion), kptfilev1.KptFileName)


for _, resource := range  resourceList.Items {
    gvk := resid.GvkFromNode(resource)
    if gvk.Equals(pkgContextGVK) {
          continue
    }
    if gvk.Equals(KptfileGVK) {
          ...
    }
}
```

File: internal/builtins/pkg_context.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+	return framework.Execute(pc, rw)
+}
+
+func (pc *PackageContextGenerator) Process(resourceList *framework.ResourceList) error {
+	var contextResources, updatedResources []*yaml.RNode
+
+	// This loop does the following:
+	// - Filters out package context resources from the input resources
+	// - Generates a package context resource for each kpt package (i.e Kptfile)
+	for _, resource := range resourceList.Items {
```
Comment (yuwenma) [inline internal/builtins/pkg_context.go:55]: would it be more readable if 

```go
var pkgContextGVK = resid.NewGvk("kptfile.kpt.dev", "v1", "ConfigMap")
var KptfileGVK =  resid.NewGvk(resid.kptfilev1.TypeMeta.APIVersion(kptfilev1.TypeMeta.APIVersion), kptfilev1.KptFileName)


for _, resource := range  resourceList.Items {
    gvk := resid.GvkFromNode(resource)
    if gvk.Equals(pkgContextGVK) {
          continue
    }
    if gvk.Equals(KptfileGVK) {
          ...
    }
}
```

File: internal/builtins/pkg_context.go (hunk 9 above)
Comment (yuwenma) [inline internal/builtins/pkg_context.go:61]: would it be more readable if 

```go
var pkgContextGVK = resid.NewGvk("kptfile.kpt.dev", "v1", "ConfigMap")
var KptfileGVK =  resid.NewGvk(resid.kptfilev1.TypeMeta.APIVersion(kptfilev1.TypeMeta.APIVersion), kptfilev1.KptFileName)


for _, resource := range  resourceList.Items {
    gvk := resid.GvkFromNode(resource)
    if gvk.Equals(pkgContextGVK) {
          continue
    }
    if gvk.Equals(KptfileGVK) {
          ...
    }
}
```

File: internal/fnruntime/runner.go [hunk 11]
```diff
@@ -74,41 +75,46 @@ func NewRunner(
...
 	if fltr.Run == nil {
-		switch {
-		case f.Image != "":
-			cfn := &ContainerFn{
-				Path:            pkgPath,
-				Image:           f.Image,
-				ImagePullPolicy: imagePullPolicy,
-				Ctx:             ctx,
-				FnResult:        fnResult,
+		if f.Image == "builtins/gen-pkg-context" {
```
Comment (yuwenma) [inline internal/fnruntime/runner.go:78]: what happens if users configure "builtins/gen-pkg-context"  in the `pipeline` field and potentially provide the flag to include meta resources to the function? (IIUC, the Kptfile resource is converted the pkgContext and excluded from the input resources.

File: internal/util/get/get.go [hunk 12]
```diff
@@ -116,6 +121,23 @@ func (c Command) Run(ctx context.Context) error {
...
 
+	if c.IsDeploymentInstance {
+		pr := printer.FromContextOrDie(ctx)
+		pr.Printf("\nCustomizing package for deployment.\n")
+		hookCmd := hook.Executor{
+			PkgPath: c.Destination,
+		}
+		builtinHooks := []kptfilev1.Function{
+			{
+				Image: "builtins/gen-pkg-context",
```
Comment (yuwenma) [inline internal/util/get/get.go:132]: nit: how about const var since the `builtins/gen-pkg-context` is referred in other places as well"?

Comment (droot) [reply internal/util/get/get.go]: Done.

**Review by droot — COMMENTED** [review]
Thanks for the review @yuwenma . Addressed the comments. PTAL.

File: internal/builtins/pkg_context.go [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	annotations := map[string]string{
+		kioutil.PathAnnotation: path.Join(path.Dir(kptfilePath), pkgContextFile),
+	}
+
+	for k, v := range annotations {
+		if _, err := cm.Pipe(yaml.SetAnnotation(k, v)); err != nil {
+			return nil, err
+		}
+	}
```
[the remark this answers is not in the document]
Comment (yuwenma) [reply internal/builtins/pkg_context.go:128]: oh, right. there's a fix but not released yet IIRC.

File: internal/fnruntime/runner.go (hunk 11 above)
Comment (yuwenma) [reply internal/fnruntime/runner.go]: Right, I'm afraid the builtin/gen-pkg-context changes the way how `--include-meta-resources` flag works unexpectedly: The `include-meta-resources` flag means some functions require the Kptfile, but  `builtins/gen-pkg-context` converts the kptfile to something else (pkg-context).  Is there a way we can prevent that from happening?

**Review by yuwenma — APPROVED** [review]
one comment about how `include-meta-resource` should behave if the `builtin/pkg-context` are unfortunately changed kptfile to pkg-context where kptfile is needed by other functions. This is more a corner case and no need to block this PR.

**Pull request merged**